### PR TITLE
Adds support for more currencies

### PR DESCRIPTION
**Pull request by jonrandy**
#### 1. Objective

Add support for the additional currencies that have been launched on the backend recently.

```
TH:

Australian Dollar (AUD)
Swiss Franc (CHF)
Chinese Yuan (CNY)
Danish Krone (DKK)
Euro (EUR)
British Pound (GBP)
Hongkong Dollar (HKD)
Japanese Yen (JPY)
Singapore Dollar (SGD)
US Dollar (USD)

SG:

Australian Dollar (AUD)
Canadian Dollar (CAD)
Swiss Franc (CHF)
China Yuan (CNY)
Euro (EUR)
British Pound (GBP)
Hongkong Dollar (HKD)
Japanese Yen (JPY)
Malaysian Ringgit (MYR)
Thai Baht (THB)
US Dollar (USD)
```

#### 2. Description of change

Modifications to the functions that:

a) Convert amounts to subunits
b) Check if a given currency is supported

#### 3. Quality assurance

Tested on my test sites for PrestaShop 1.6 and 1.7 - attempting payments in the new currencies

#### 4. Impact of the change

Newly added currencies can be used

#### 5. Priority of change

Normal

#### 6. Additional notes

👹

### CODE REVIEW

File: omise/libraries/omise-plugin/helpers/charge.php [hunk 1]
```diff
@@ -2,25 +2,73 @@
...
+                "GNF" => 0,
+                "ISK" => 0,
+                "JPY" => 0,
+                "KMF" => 0,
+                "KRW" => 0,
+                "PYG" => 0,
+                "RWF" => 0,
+                "UGX" => 0,
+                "UYI" => 0,
+                "VND" => 0,
```
Comment (jacstn) [inline omise/libraries/omise-plugin/helpers/charge.php:20]: why all those currencies are listed since they are not supported?

Comment (jonrandy) [reply omise/libraries/omise-plugin/helpers/charge.php:20]: The list does not relate to supported currencies. It is used in a generic function to get the number of subunits for a given currency. I just added those so the function is very future proof

Comment (jacstn) [reply omise/libraries/omise-plugin/helpers/charge.php:20]: why no `PLN` polish złoty, why no `Russian Rubel`? But `Uganda Shilling` is there?

**Comment by jacstn** [comment]
@guzzilar, @jonrandy I am implementing that in `mag2`
different currencies are supported by different account country origin.
any ideas on how to solve that?

File: omise/libraries/omise-plugin/helpers/charge.php (hunk 1 above)
Comment (jonrandy) [reply omise/libraries/omise-plugin/helpers/charge.php:20]: Explained in the code. Doesn't need to be there if it has 2 decimal places

Comment (jacstn) [reply omise/libraries/omise-plugin/helpers/charge.php:20]: ok, so what it does, it handles all currencies in the world, right?

Comment (jonrandy) [reply omise/libraries/omise-plugin/helpers/charge.php:20]: Assuming the list I pulled the data from was correct - yes - at least until new currencies are added, but even then the likelihood is the new currency will probably have 2 decimal places like the majority of others

Comment (jacstn) [reply omise/libraries/omise-plugin/helpers/charge.php:20]: ok... what are the chances that those currencies will be added by omise like UGX,
if you want to leave that in the code ok.
But that most likely will never, ever be used, ever...

**Review by jacstn — APPROVED** [review]

File: omise/libraries/omise-plugin/helpers/charge.php [hunk 2]
```diff
@@ -2,25 +2,73 @@
...
+        public static function subUnitsFor($curr) {
+            $c = strtoupper($curr);
+            $decimals = isset(self::$currencyDecimals[$c]) ? self::$currencyDecimals[$c] : self::$defaultCurrencyDecimals;
+            return pow(10, $decimals);
+        }        
+
         /**
          *
          * @param string $currency
-         * @param integer $amount
```
Comment (jacstn) [inline omise/libraries/omise-plugin/helpers/charge.php:8]: we could correct comment maybe `@param float $amount` ?

File: omise/libraries/omise-plugin/helpers/charge.php (hunk 1 above)
Comment (jonrandy) [reply omise/libraries/omise-plugin/helpers/charge.php:20]: Yeah, I was conflicted about whether to put them all in, but eventually decided it was better to have the function work properly for as many cases as possible without being wrong, or having to return something like `null` for a currency it didn't know. You're correct though, most are edge cases that will almost certainly never be used

File: omise/libraries/omise-plugin/helpers/charge.php (hunk 2 above)
Comment (jonrandy) [reply omise/libraries/omise-plugin/helpers/charge.php:8]: It accepts any var that can be 'used' as a number - I guess that would be mainly int/float/numeric string - maybe should just write `@param numeric $amount`

**Comment by jonrandy** [comment]
> @guzzilar, @jonrandy I am implementing that in `mag2`
> different currencies are supported by different account country origin.
> any ideas on how to solve that?

Ultimately we need to use the capability API for it I think - each payment method lists the currencies it can use. One would assume these will be accurate

**Comment by jacstn** [comment]
will not happen anytime soon - getting data in capabilities.

**Comment by jonrandy** [comment]
It's already there. PR is awaiting merge

**Pull request merged**